Omnitrace docs refactoring (#353)
ROCm/omnitrace: docs/how-to/sampling-call-stack.rst

Sampling the call stack
****************************************************

`Omnitrace <https://github.com/ROCm/omnitrace>`_  call-stack sampling can be activated 
with either a binary instrumented by the ``omnitrace`` executable 
or via the ``omnitrace-sample`` executable.
All of the commands below are effectively equivalent:

* Binary rewrite with only instrumentation necessary to start/stop sampling

  .. code-block:: shell

     omnitrace-instrument -M sampling -o foo.inst -- foo
     omnitrace-run -- ./foo.inst

* Runtime instrumentation with only instrumentation necessary to start/stop sampling

  .. code-block:: shell

     omnitrace-instrument -M sampling -- foo

* No instrumentation required

  .. code-block:: shell

     omnitrace-sample -- foo

.. note::

   Set ``OMNITRACE_USE_SAMPLING=ON`` to activate call-stack sampling when executing an instrumented binary.

All ``omnitrace-instrument -M sampling`` (referred to as "instrumented-sampling" henceforth) 
does is wrap the ``main`` of the executable with initialization
before ``main`` starts and finalization after ``main`` ends.
This can be easily accomplished without instrumentation via a ``LD_PRELOAD`` 

> of a library with containing a dynamic symbol wrapper around ```__libc_start_main``.

Reviewer: ```suggestion
of a library with containing a dynamic symbol wrapper around ``__libc_start_main``.
```

or by using the ``omnitrace-sample`` executable.
All of the following commands are effectively equivalent:

* Binary rewrite with only the instrumentation necessary to start and stop sampling

  .. code-block:: shell

     omnitrace-instrument -M sampling -o foo.inst -- foo
     omnitrace-run -- ./foo.inst

* Runtime instrumentation with only the instrumentation necessary to start and stop sampling

  .. code-block:: shell

     omnitrace-instrument -M sampling -- foo

* No instrumentation required

  .. code-block:: shell

     omnitrace-sample -- foo

.. note::

   Set ``OMNITRACE_USE_SAMPLING=ON`` to activate call-stack sampling when executing an instrumented binary.

All ``omnitrace-instrument -M sampling`` (subsequently referred to as "instrumented-sampling") 
does is wrap the ``main`` of the executable with initialization
before ``main`` starts and finalization after ``main`` ends.
This can be accomplished without instrumentation through a ``LD_PRELOAD`` 
of a library containing a dynamic symbol wrapper around ``__libc_start_main``.
As a result, whenever binary instrumentation is unnecessary, the use of ``omnitrace-sample`` 

> is recommended over ``omnitrace-instrument -M sampling`` for several reasons:

Author: Sure, that makes sense. We won't need the comma then.

* ``omnitrace-sample`` provides command-line options for controlling the Omnitrace feature set instead of 
  requiring configuration files or environment variables
* Despite the fact that instrumented-sampling only requires inserting snippets 
  around one function (``main``), Dyninst
  does not have a feature for specifying that parsing and processing all the 
  other symbols in the binary is unnecessary.
  In the best-case scenario when the target binary is relatively small, 
  instrumented-sampling has a slightly slower launch time,
  but in the worst case scenarios it requires a significant amount of time and memory to launch.
* ``omnitrace-sample`` is fully compatible with MPI, for example in the command ``mpirun -n 2 omnitrace-sample -- foo``, 
  whereas ``mpirun -n 2 omnitrace-instrument -M sampling -- foo``
  is incompatible with some MPI distributions (particularly OpenMPI) because of 

>   MPI restrictions against forking within an MPI rank

Author: Yes, I think this is better organized.

All arguments preceding the double hyphen
are interpreted as belonging to Omnitrace and all arguments following it 
are interpreted as the
application and its arguments. The double hyphen is only necessary when passing 
command-line arguments to a target
which also uses hyphens. For example, you can run ``omnitrace-sample ls``, but 
to run ``ls -la``, use ``omnitrace-sample -- ls -la``.

:doc:`Configuring the Omnitrace runtime options <./configuring-runtime-options>` 
establishes the precedence of environment variable values over values specified 
in the configuration files. This enables
you to configure the Omnitrace runtime to your preferred default behavior 
in a file such as ``~/.omnitrace.cfg`` and then easily override
those settings using a command like ``OMNITRACE_ENABLED=OFF omnitrace-sample -- foo``.
Similarly, the command-line arguments passed to ``omnitrace-sample`` take precedence 
over environment variables.

All of the command-line options above correlate to one or more configuration 
settings, for example, ``--cpu-events`` correlates to the ``OMNITRACE_PAPI_EVENTS`` configuration variable.

> After the command-line arguments to ``omnitrace-sample`` have been processed but 
> before the target application runs, ``omnitrace-sample`` creates a log
> showing which environment variables were set or modified:

Reviewer 2: ```suggestion
``omnitrace-sample`` will process the arguments and output a summary of its configuration before the target application is executed. 
The following few snippets show ``omnitrace-sample`` run with various environment updates.
```
Maybe then, the following 3 snippets can be presented as an unordered list to visualize that they're all part of the same "idea".

Author: I made a few additional adjustments to make this work as a list.

   ...

The snippet below shows the environment updates when ``omnitrace-sample`` enables 
profiling, tracing, host process-sampling, and device process-sampling,
sets the output path to ``omnitrace-output`` and the output prefix to ``%tag%``, and disables 
all the available backends:

.. code-block:: shell

   $ omnitrace-sample -PTDH -E all -o omnitrace-output %tag% -- ./parallel-overhead-locks 30 4 100

   LD_PRELOAD=/opt/omnitrace/lib/libomnitrace-dl.so.1.7.1
   OMNITRACE_CPU_FREQ_ENABLED=true
   OMNITRACE_OUTPUT_PATH=omnitrace-output
   OMNITRACE_OUTPUT_PREFIX=%tag%
   OMNITRACE_TRACE_THREAD_LOCKS=false
   OMNITRACE_TRACE_THREAD_RW_LOCKS=false
   OMNITRACE_TRACE_THREAD_SPIN_LOCKS=false
   OMNITRACE_USE_KOKKOSP=false
   OMNITRACE_USE_MPIP=false
   OMNITRACE_USE_OMPT=false
   OMNITRACE_TRACE=true
   OMNITRACE_USE_PROCESS_SAMPLING=true
   OMNITRACE_USE_RCCLP=false
   OMNITRACE_USE_ROCM_SMI=false
   OMNITRACE_USE_ROCPROFILER=false
   OMNITRACE_USE_ROCTRACER=false
   OMNITRACE_USE_ROCTX=false
   OMNITRACE_USE_SAMPLING=true
   OMNITRACE_PROFILE=true
   ...

An ``omnitrace-sample`` example
========================================

> 

Reviewer 2: Below is the full output from the previously described example. 

You can probably word this better than me. :) I'm just giving this example a little bit of an intro. It is the same command-line as the example above;  ie. ` $ omnitrace-sample -PTDH -E all -o omnitrace-output %tag% -- ./parallel-overhead-locks 30 4 100`. This is just the full output.

Author: Sure. I'll work on this tomorrow and will get your input on the results. Thanks for all the feedback!

.. code-block:: shell

   $ omnitrace-sample -PTDH -E all -o omnitrace-output %tag% -c -- ./parallel-overhead-locks 30 4 100

   LD_PRELOAD=/opt/omnitrace/lib/libomnitrace-dl.so.1.7.1
   OMNITRACE_CONFIG_FILE=
   OMNITRACE_CPU_FREQ_ENABLED=true
   OMNITRACE_OUTPUT_PATH=omnitrace-output
   OMNITRACE_OUTPUT_PREFIX=%tag%
   OMNITRACE_TRACE_THREAD_LOCKS=false
   OMNITRACE_TRACE_THREAD_RW_LOCKS=false
   OMNITRACE_TRACE_THREAD_SPIN_LOCKS=false

>    OMNITRACE_USE_KOKKOSP=false
>    OMNITRACE_USE_MPIP=false
>    OMNITRACE_USE_OMPT=false
>    OMNITRACE_TRACE=true
>    OMNITRACE_USE_PROCESS_SAMPLING=true
>    OMNITRACE_USE_RCCLP=false
>    OMNITRACE_USE_ROCM_SMI=false
>    OMNITRACE_USE_ROCPROFILER=false
>    OMNITRACE_USE_ROCTRACER=false
>    OMNITRACE_USE_ROCTX=false
>    OMNITRACE_USE_SAMPLING=true
>    OMNITRACE_PROFILE=true
> 
>    [omnitrace][omnitrace_init_tooling] Instrumentation mode: Sampling
> 
> 
>          ______   .___  ___. .__   __.  __  .___________..______          ___       ______  _______
>       /  __  \  |   \/   | |  \ |  | |  | |           ||   _  \        /   \     /      ||   ____|
>       |  |  |  | |  \  /  | |   \|  | |  | `---|  |----`|  |_)  |      /  ^  \   |  ,----'|  |__
>       |  |  |  | |  |\/|  | |  . `  | |  |     |  |     |      /      /  /_\  \  |  |     |   __|
>       |  `--'  | |  |  |  | |  |\   | |  |     |  |     |  |\  \----./  _____  \ |  `----.|  |____
>       \______/  |__|  |__| |__| \__| |__|     |__|     | _| `._____/__/     \__\ \______||_______|
> 
> 
>    [759.689]       perfetto.cc:55903 Configured tracing session 1, #sources:1, duration:0 ms, #buffers:1, total buffer size:1024000 KB, total sessions:1, uid:0 session name: ""
> 
>    [parallel-overhead-locks] Threads: 4
>    [parallel-overhead-locks] Iterations: 100
>    [parallel-overhead-locks] fibonacci(30)...
>    [1] number of iterations: 100
>    [2] number of iterations: 100
>    [3] number of iterations: 100
>    [4] number of iterations: 100
>    [parallel-overhead-locks] fibonacci(30) x 4 = 394644873
>    [parallel-overhead-locks] number of mutex locks = 400
>    [omnitrace][107157][0][omnitrace_finalize]
>    [omnitrace][107157][0][omnitrace_finalize] finalizing...
>    [omnitrace][107157][0][omnitrace_finalize]
>    [omnitrace][107157][0][omnitrace_finalize] omnitrace/process/107157 : 0.610427 sec wall_clock,    2.248 MB peak_rss,    2.265 MB page_rss, 2.560000 sec cpu_clock,  419.4 % cpu_util [laps: 1]
>    [omnitrace][107157][0][omnitrace_finalize] omnitrace/process/107157/thread/0 : 0.608866 sec wall_clock, 0.000677 sec thread_cpu_clock,    0.1 % thread_cpu_util,    2.248 MB peak_rss [laps: 1]
>    [omnitrace][107157][0][omnitrace_finalize] omnitrace/process/107157/thread/1 : 0.608237 sec wall_clock, 0.603553 sec thread_cpu_clock,   99.2 % thread_cpu_util,    2.204 MB peak_rss [laps: 1]
>    [omnitrace][107157][0][omnitrace_finalize] omnitrace/process/107157/thread/2 : 0.601430 sec wall_clock, 0.598378 sec thread_cpu_clock,   99.5 % thread_cpu_util,    1.156 MB peak_rss [laps: 1]
>    [omnitrace][107157][0][omnitrace_finalize] omnitrace/process/107157/thread/3 : 0.570223 sec wall_clock, 0.568713 sec thread_cpu_clock,   99.7 % thread_cpu_util,    0.772 MB peak_rss [laps: 1]
>    [omnitrace][107157][0][omnitrace_finalize] omnitrace/process/107157/thread/4 : 0.557637 sec wall_clock, 0.557198 sec thread_cpu_clock,   99.9 % thread_cpu_util,    0.156 MB peak_rss [laps: 1]
>    [omnitrace][107157][0][omnitrace_finalize]
>    [omnitrace][107157][0][omnitrace_finalize] Finalizing perfetto...
>    [omnitrace][107157][perfetto]> Outputting '/home/user/data/omnitrace-output/2022-10-19_02.46/parallel-overhead-locksperfetto-trace-107157.proto' (842.90 KB / 0.84 MB / 0.00 GB)... Done
>    [omnitrace][107157][trip_count]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockstrip_count-107157.json'
>    [omnitrace][107157][trip_count]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockstrip_count-107157.txt'
>    [omnitrace][107157][sampling_percent]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockssampling_percent-107157.json'
>    [omnitrace][107157][sampling_percent]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockssampling_percent-107157.txt'
>    [omnitrace][107157][sampling_cpu_clock]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockssampling_cpu_clock-107157.json'
>    [omnitrace][107157][sampling_cpu_clock]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockssampling_cpu_clock-107157.txt'
>    [omnitrace][107157][sampling_wall_clock]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockssampling_wall_clock-107157.json'
>    [omnitrace][107157][sampling_wall_clock]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockssampling_wall_clock-107157.txt'
>    [omnitrace][107157][wall_clock]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockswall_clock-107157.json'
>    [omnitrace][107157][wall_clock]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-lockswall_clock-107157.txt'
>    [omnitrace][107157][metadata]> Outputting 'omnitrace-output/2022-10-19_02.46/parallel-overhead-locksmetadata-107157.json' and 'omnitrace-output/2022-10-19_02.46/parallel-overhead-locksfunctions-107157.json'
>    [omnitrace][107157][0][omnitrace_finalize] Finalized
>    [761.584]       perfetto.cc:57382 Tracing session 1 ended, total sessions:0

Reviewer 2: Just updating output using the latest Omnitrace version. 

```suggestion
$ omnitrace-sample -PTDH -E all -o omnitrace-output %tag% -c -- ./parallel-overhead-locks 30 4 100

LD_PRELOAD=/home/gliff/code/omnitrace/build-release/lib/libomnitrace-dl.so.1.11.3
OMNITRACE_CONFIG_FILE=
OMNITRACE_CPU_FREQ_ENABLED=true
OMNITRACE_OUTPUT_PATH=omnitrace-output
OMNITRACE_OUTPUT_PREFIX=%tag%
OMNITRACE_PROFILE=true
OMNITRACE_TRACE=true
OMNITRACE_TRACE_THREAD_LOCKS=false
OMNITRACE_TRACE_THREAD_RW_LOCKS=false
OMNITRACE_TRACE_THREAD_SPIN_LOCKS=false
OMNITRACE_USE_KOKKOSP=false
OMNITRACE_USE_MPIP=false
OMNITRACE_USE_OMPT=false
OMNITRACE_USE_PROCESS_SAMPLING=true
OMNITRACE_USE_RCCLP=false
OMNITRACE_USE_ROCM_SMI=false
OMNITRACE_USE_ROCPROFILER=false
OMNITRACE_USE_ROCTRACER=false
OMNITRACE_USE_ROCTX=false
OMNITRACE_USE_SAMPLING=true

[omnitrace][dl][1785877] omnitrace_main
[omnitrace][1785877][omnitrace_init_tooling] Instrumentation mode: Sampling

      ______   .___  ___. .__   __.  __  .___________..______          ___       ______  _______
     /  __  \  |   \/   | |  \ |  | |  | |           ||   _  \        /   \     /      ||   ____|
    |  |  |  | |  \  /  | |   \|  | |  | `---|  |----`|  |_)  |      /  ^  \   |  ,----'|  |__
    |  |  |  | |  |\/|  | |  . `  | |  |     |  |     |      /      /  /_\  \  |  |     |   __|
    |  `--'  | |  |  |  | |  |\   | |  |     |  |     |  |\  \----./  _____  \ |  `----.|  |____
     \______/  |__|  |__| |__| \__| |__|     |__|     | _| `._____/__/     \__\ \______||_______|

    omnitrace v1.11.2 (rev: 2586b74db8bf335742600010b8d9f1ce8da9cf89, compiler: GNU v11.4.1, rocm: v6.1.x)
[988.958]       perfetto.cc:58649 Configured tracing session 1, #sources:1, duration:0 ms, #buffers:1, total buffer size:1024000 KB, total sessions:1, uid:0 session name: ""

[parallel-overhead-locks] Threads: 4
[parallel-overhead-locks] Iterations: 100
[parallel-overhead-locks] fibonacci(30)...
[1] number of iterations: 100
[2] number of iterations: 100
[3] number of iterations: 100
[4] number of iterations: 100
[parallel-overhead-locks] fibonacci(30) x 4 = 409221992
[parallel-overhead-locks] number of mutex locks = 400

[omnitrace][1785877][0][omnitrace_finalize] finalizing...
[omnitrace][1785877][0][omnitrace_finalize] 
[omnitrace][1785877][0][omnitrace_finalize] omnitrace/process/1785877 : 0.294342 sec wall_clock,    4.776 MB peak_rss,    3.170 MB page_rss, 0.990000 sec cpu_clock,  336.3 % cpu_util [laps: 1]
[omnitrace][1785877][0][omnitrace_finalize] omnitrace/process/1785877/thread/0 : 0.291535 sec wall_clock, 0.002619 sec thread_cpu_clock,    0.9 % thread_cpu_util,    4.776 MB peak_rss [laps: 1]
[omnitrace][1785877][0][omnitrace_finalize] omnitrace/process/1785877/thread/1 : 0.271353 sec wall_clock, 0.222572 sec thread_cpu_clock,   82.0 % thread_cpu_util,    4.200 MB peak_rss [laps: 1]
[omnitrace][1785877][0][omnitrace_finalize] omnitrace/process/1785877/thread/2 : 0.238218 sec wall_clock, 0.206405 sec thread_cpu_clock,   86.6 % thread_cpu_util,    3.432 MB peak_rss [laps: 1]
[omnitrace][1785877][0][omnitrace_finalize] omnitrace/process/1785877/thread/3 : 0.209459 sec wall_clock, 0.193415 sec thread_cpu_clock,   92.3 % thread_cpu_util,    2.472 MB peak_rss [laps: 1]
[omnitrace][1785877][0][omnitrace_finalize] omnitrace/process/1785877/thread/4 : 0.212029 sec wall_clock, 0.211694 sec thread_cpu_clock,   99.8 % thread_cpu_util,    1.152 MB peak_rss [laps: 1]
[omnitrace][1785877][0][omnitrace_finalize] 
[omnitrace][1785877][0][omnitrace_finalize] Finalizing perfetto...
[omnitrace][1785877][perfetto]> Outputting '/home/gliff/code/omnitrace/build-release/omnitrace-output/2024-07-15_16.21/parallel-overhead-locksperfetto-trace-1785877.proto' (39.12 KB / 0.04 MB / 0.00 GB)... Done
[omnitrace][1785877][wall_clock]> Outputting 'omnitrace-output/2024-07-15_16.21/parallel-overhead-lockswall_clock-1785877.json'
[omnitrace][1785877][wall_clock]> Outputting 'omnitrace-output/2024-07-15_16.21/parallel-overhead-lockswall_clock-1785877.txt'
[omnitrace][1785877][metadata]> Outputting 'omnitrace-output/2024-07-15_16.21/parallel-overhead-locksmetadata-1785877.json' and 'omnitrace-output/2024-07-15_16.21/parallel-overhead-locksfunctions-1785877.json'
[omnitrace][1785877][0][omnitrace_finalize] Finalized: 0.054582 sec wall_clock,    0.000 MB peak_rss,   -1.798 MB page_rss, 0.040000 sec cpu_clock,   73.3 % cpu_util
[989.312]       perfetto.cc:60128 Tracing session 1 ended, total sessions:0
```

are interpreted as the
application and its arguments. The double hyphen is only necessary when passing 
command-line arguments to a target
which also uses hyphens. For example, you can run ``omnitrace-sample ls``, but 
to run ``ls -la``, use ``omnitrace-sample -- ls -la``.

:doc:`Configuring the Omnitrace runtime options <./configuring-runtime-options>` 
establishes the precedence of environment variable values over values specified 
in the configuration files. This enables
you to configure the Omnitrace runtime to your preferred default behavior 
in a file such as ``~/.omnitrace.cfg`` and then easily override
those settings in the command line, for example, ``OMNITRACE_ENABLED=OFF omnitrace-sample -- foo``.
Similarly, the command-line arguments passed to ``omnitrace-sample`` take precedence 
over environment variables.

All of the command-line options above correlate to one or more configuration 
settings, for example, ``--cpu-events`` correlates to the ``OMNITRACE_PAPI_EVENTS`` configuration variable.
``omnitrace-sample`` processes the arguments and outputs a summary of its configuration 
before running the target application. 

> The following snippets show how ``omnitrace-sample`` runs with various environment updates.

Author: @dgaliffiAMD This is the part I reformatted based on your feedback. Please let me know if it looks good now.
